### PR TITLE
Selectable disallow (metadata disallow)

### LINKED ISSUE #360: Allow select for major or minor for security
**Describe the bug/issue**
Currently we check by default if a major bump in version happened to disable update.
This behavior is used for native code change to prevent install update on device who doesn’t have the native code.
Over time I saw many complaints about doing a major bump for something final users don’t care.
Even if the display version can be different in App Store it seems annoying.

So one change who could be nice is to allow capgo user to set how disable work.
- major 
- minor
- version number 
- None

So The current behavior is “major”, minor will be check if the minor coming from device ( 2 in 1.2.3) is equal or above the latest in the channel.

version number will use the min version send as metadata in the upload step( cli should be modified as well), this version has to be displayed in the UI, if this data is missing and this setting is active, it will use it’s own number as min number.

an none will be no check, same as we disable the feature currently.

this settings should be set from the UI or the CLI.

By default the behavior should be major, and migration script should be provided to migrate all users.

### PR DESCRIPTION
**Pull request by WcaleNieWolny**
This PR fixes #360. /claim #360

First of all here is a brief list of changes:
 * Playwright + supabase integration
 * Improve playwright auth
 * Show an error if the channel is misconfigured (metadata is lacking)
image
 * Show the minimal update version if the metadata disallow is on in both channel and bundle
 * Add a switch to select update strategy in channel settings
 * Add a CLI change to make sure that if updates are disabled based on this metadata then require this metadata to be present when uploading (different PR)
 * Add minor disallow (if updating from `1.2.0` -> `1.3.0` this option will prevent such update)
 * Change backend to check metadata and disallow update based on that
 * Update playwright
 * Add tests for all of the features I mentioned (fronend, backend + CLI)
 
At this time I will not be providing a video of the changes as it would be near impossible to show all edge cases in that video. I believe the extensive tests for this feature are enough.

### CODE REVIEW

**Comment by WcaleNieWolny** [comment]
The tests failed, I will rerun them as I changed some things

/test all

**Comment by WcaleNieWolny** [comment]
The new tests are failing because of the state of the db when the tests are being run. I have only tested this partially (I haven never ran `all` test before submitting this PR) I am working on improving this.

@ologbonowiwi could you review this PR if you have a sec

**Comment by WcaleNieWolny** [comment]
I have debugged and improved the tests. I think I remove the weird state issues within the new tests.

/test all

File: locales/en.yml
Comment (wesleymatosdev) [inline locales/en.yml:1]: nice changes, re-organizing the order of the keys.

one thing, though, `yes` and `no` are the only ones with a single comma (`'`) on the key; is that intentional?

File: src/components/tables/ChannelTable.vue [hunk 1]
```diff
@@ -142,6 +159,28 @@ async function getData() {
...
+      if (channel.version.minUpdateVersion === null) {
+        channel.misconfigured = true
+        anyMisconfigured = true
+      }
+
+      if ((channel.enable_progressive_deploy || channel.enableAbTesting) && channel.secondVersion.minUpdateVersion === null) {
+        channel.misconfigured = true
+        anyMisconfigured = true
+      }
+    }
```
Comment (wesleymatosdev) [inline src/components/tables/ChannelTable.vue:180]: you could use the `.map` instead of looping through the array twice

something like:
```suggestion
    .map(e => {
      const channel = e as any as typeof element

      if (channel.version.minUpdateVersion === null) {
        channel.misconfigured = true
        anyMisconfigured = true
      }
      if ((channel.enable_progressive_deploy || channel.enableAbTesting) && channel.secondVersion.minUpdateVersion === null) {
        channel.misconfigured = true
        anyMisconfigured = true
      }
      
      return channel
    })
```

Comment (wesleymatosdev) [reply src/components/tables/ChannelTable.vue:180]: when we loop it only once, one thing that we could improve is the conditionals.

we can use names to explain the conditions and set misconfigured only once

an example: 
```typescript
      const minUpdateVersionIsNull = channel.version.minUpdateVersion === null
      const isProgressiveDeployOrAbTesting = channel.enable_progressive_deploy || channel.enableAbTesting
      const secondMinUpdateVersionIsNull = channel.secondVersion.minUpdateVersion === null
      const nameToBeChanged = isProgressiveDeployOrAbTesting && secondMinUpdateVersionIsNull

      if (minUpdateVersionIsNull || nameToBeChanged) {
          channel.misconfigured = true
          anyMisconfigured = true
      }

     return channel
```

File: src/pages/app/p/[p]/bundle/[bundle].vue
Comment (wesleymatosdev) [inline src/pages/app/p/[p]/bundle/[bundle].vue:1]: there are two console logs (lines 69 and 453); these will be merged, or are they only debug consoles?

File: tests/auth.setup.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { expect, test as setup } from '@playwright/test'
```
Comment (wesleymatosdev) [inline tests/auth.setup.ts:1]: this can bring confusion since it's well spread by the node community to have a `test` key that declares test cases

when I saw `setup` on the code, I went back to the imports to see if it was a method we made and not from the test library. this likely will bring confusion to contributors

```suggestion
import { expect, test } from '@playwright/test'
```

File: tests/auth.setup.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  await page.getByRole('button', { name: 'Log in' }).click()
+
+  // Expect the URL to change to the logged in dashboard
+  await expect(page).toHaveURL(`${BASE_URL}/app/home`, { timeout: 60_000 })
+
+  await page.context().storageState({ path: authFile1 })
+})
+
+// This will be usefull some day
+// setup('authenticate as test2', async ({ page }) => {
```
Comment (wesleymatosdev) [inline tests/auth.setup.ts:24]: Since setup is actually `test` from the playwright, you can uncomment this code and use `test.skip`

File: tests/auth.setup.ts
Comment (wesleymatosdev) [inline tests/auth.setup.ts:1]: why this file is called `.setup` rather than `.test` our `.spec`?

File: tests/selectable_disallow.spec.ts
Comment (wesleymatosdev) [inline tests/selectable_disallow.spec.ts:1]: rather than using css locators (`something > other_thing`), you can set data-testid on the component and use the [getByTestId playwright's locator](https://playwright.dev/docs/locators#locate-by-test-id) instead.

It's more reliable and uncoupled since it doesn't depend on how the HTML is structured

File: tests/utils.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    return supaClient
+  // eslint-disable-next-line n/prefer-global/process
+  supaClient = createClient<Database>(process.env.SUPABASE_URL ?? defaultSupabaseUrl, process.env.SUPABASE_ANON ?? defaultSupabaseAnonKey, options)
+
+  const { error } = await supaClient.auth.signInWithPassword({
+    email: DEFAULT_EMAIL,
+    password: DEFAULT_PASSWORD,
+  })
+
+  expect(error).toBeNull()
```
Comment (wesleymatosdev) [inline tests/utils.ts:42]: Excellent use of expectation here to handle the error!

File: tests_backend/run_backend_tests.ts [hunk 5]
```diff
@@ -24,7 +24,7 @@ let backendType: 'none' | 'local' | 'upstash' | null = null
 const noRedisEnvFilePath = await getEnvFile('none')
 const localRedisEnvFilePath = await getEnvFile('local')
 
-async function getAdminSupabaseTokens(): Promise<{ url: string; serviceKey: string }> {
+async function getAdminSupabaseTokens(): Promise<{ url: string; serviceKey: string; anonToken: string }> {
```
Comment (wesleymatosdev) [inline tests_backend/run_backend_tests.ts:27]: we could give a better name than `anonToken` here, something like `anonymousToken`, `publicToken`, etc.; something that auto-explains what this thing is

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]
Left some comments but the core stuff looks great, and it's nice that we're adding more tests almost on each PR 🚀

**Comment by WcaleNieWolny** [comment]
I really am trying to reproduce this CI fail, I have change the playwright config. I will test again all and selectable_disallow without the rest. I am not sure why this if failing, I cannot reproduce this fail in my local setup

File: tests_backend/run_backend_tests.ts (hunk 5 above)
Comment (WcaleNieWolny) [reply tests_backend/run_backend_tests.ts:27]: This is called `supa_anon` in config.js `anon key` in `supabase status`. Anon is the term used officially by supabase

File: tests/utils.ts (hunk 4 above)
Comment (WcaleNieWolny) [reply tests/utils.ts:42]: thanks 😄

File: locales/en.yml
Comment (WcaleNieWolny) [reply locales/en.yml:1]: I think that the VS code extensions changed something in a weird way. I could fix that, tho it does work as it

File: src/pages/app/p/[p]/bundle/[bundle].vue
Comment (WcaleNieWolny) [reply src/pages/app/p/[p]/bundle/[bundle].vue:1]: Good catch, removed in the latest commit

File: tests/auth.setup.ts
Comment (WcaleNieWolny) [reply tests/auth.setup.ts:1]: I have followed the [official tutorial](https://playwright.dev/docs/auth) for this. This is not here to test auth, rather to remove the requirement for each invididual test to authenticate. This is here to prevent code duplication and not to test anything

File: tests/auth.setup.ts (hunk 2 above)
Comment (WcaleNieWolny) [reply tests/auth.setup.ts:1]: Again, I have used the [official tutorial(https://playwright.dev/docs/auth) and that is how it has been done there

File: tests/auth.setup.ts (hunk 3 above)
Comment (WcaleNieWolny) [reply tests/auth.setup.ts:24]: I could also remove it. This is here only because I originally  started to work on this file for the [organization pr](https://github.com/Cap-go/capgo/pull/335/) and there having 2 users was required.

This is here so that if anyone wants to add a second user in the tests one day he or she can just uncoment this

File: tests/selectable_disallow.spec.ts
Comment (WcaleNieWolny) [reply tests/selectable_disallow.spec.ts:1]: That is an interesting thought, I will look into this

**Comment by WcaleNieWolny** [comment]
I am doing the test again. Slowly starting to think that the CI is just slower then my PC causing this to fail

/test selectable_disallow

**Comment by WcaleNieWolny** [comment]
Tests are still failing. I am wondering if this is due to the spec diff between CI/CD and my PC. I have increased the timeout. If it works in the CI/CD then I will think about improving this

/test selectable_disallow

**Comment by WcaleNieWolny** [comment]
> Test timeout of 20000ms exceeded

yeah, this is just slow in the CI/CD

File: tests_backend/run_backend_tests.ts (hunk 5 above)
Comment (wesleymatosdev) [reply tests_backend/run_backend_tests.ts:27]: Thanks for clarifying!

File: tests/auth.setup.ts
Comment (wesleymatosdev) [reply tests/auth.setup.ts:1]: Oh, I never saw it! I always used `beforeAll` to do this, but knowing another way is good.

No change on this file is needed then, thanks!

**Comment by WcaleNieWolny** [comment]
I think it could be fixed in the latest commit, I have change the state problems

/test selectable_disallow

**Comment by WcaleNieWolny** [comment]
I have changed the dev server in tests. I think I figured out why my tests failed.

/test selectable_disallow

**Comment by WcaleNieWolny** [comment]
FINALY!! After 3 days the tests passed 🥳 

/test all

**Comment by WcaleNieWolny** [comment]
Another day, another attempt. PLEASE just work

/test selectable_disallow

**Comment by WcaleNieWolny** [comment]
PERHAPS THIS WORKS NOW

/test all

**Comment by WcaleNieWolny** [comment]
I have added a fix for a /updates endpoint bug. This bug was not related to my changes

/test all

**Comment by WcaleNieWolny** [comment]
Added a test for the bug fix + fixed the endpoint

/test backend

**Comment by riderx** [comment]
thanks @ologbonowiwi  /tip $20 @ologbonowiwi
